parser/parser.go

```diff
@@ -0,0 +1,187 @@
+	"go.mongodb.org/mongo-driver/mongo/readpref"
+)
+
+type Class struct {
+	subject           string
+	catalogNumber     string
+	section           string
+	gradeDistribution bson.A
+}
+
+func getCollection(client *mongo.Client, database string, collection string) (returnCollection *mongo.Collection) {
+	return (client.Database(database).Collection(collection))
+}
+
+func DBConnect(URI string) (client *mongo.Client) {
+	client, err := mongo.Connect(context.TODO(), options.Client().ApplyURI(URI))
+	if err != nil {
+		fmt.Println(err)
+		panic(err)
+	}
+	if err := client.Ping(context.TODO(), readpref.Primary()); err != nil {
+		fmt.Println(err)
+		panic(err)
+	}
+	return client
+}
+
+func csvToClassesSlice(csvFile *os.File, logFile *os.File) (classes []Class) {
+	reader := csv.NewReader(csvFile)
+	records, err := reader.ReadAll() // records is [][]strings
+	if err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+	// look for the subject column and w column
+	subjectCol := -1
+	wCol := -1
+	for j := 0; j < len(records[0]); j++ {
+		if strings.Compare(records[0][j], "Subject") == 0 {
+			subjectCol = j
+		}
+		if strings.Compare(records[0][j], "W") == 0 || strings.Compare(records[0][j], "Total W") == 0 || strings.Compare(records[0][j], "W Total") == 0 {
+			wCol = j
+		}
+		if wCol == -1 || subjectCol == -1 {
+			continue
+		} else {
+			break
+		}
+	}
+	if wCol == -1 {
+		logFile.WriteString("could not find W column")
+	}
+	catalogNumberCol := subjectCol + 1
+	sectionCol := subjectCol + 2
+
+	for i := 1; i < len(records); i++ {
+		// convert grade distribution from string to int
+		var tempSlice bson.A
+		for j := 0; j < 13; j++ {
+			var tempInt int
+			fmt.Sscan(records[i][3+subjectCol+j], &tempInt)
+			tempSlice = append(tempSlice, tempInt)
+		}
+		// add w number to the grade_distribution slice
+		var tempInt int
+		if wCol != -1 {
+			fmt.Sscan(records[i][wCol], &tempInt)
+		}
+		tempSlice = append(tempSlice, tempInt)
+		// add new class to classes slice
+		classes = append(classes,
+			Class{
+				subject:           records[i][subjectCol],
+				catalogNumber:     records[i][catalogNumberCol],
+				section:           records[i][sectionCol],
+				gradeDistribution: tempSlice,
+			})
+	}
+	return classes
+}
+
+// inserts grades into mongodb
+func insertGrades(sectionsCollection *mongo.Collection, coursesCollection *mongo.Collection, classes []Class, academicSession string, logFile *os.File) {
+	for i := 0; i < len(classes); i++ {
+		var result bson.D
+		var err error
+		err = coursesCollection.FindOne(context.TODO(), bson.D{{"course_number", classes[i].catalogNumber}, {"subject_prefix", classes[i].subject}}).Decode(&result)
+		// if class is not in courses section
+		if err != nil {
+			// log that class could not be found
+			logFile.WriteString("could not find course " + classes[i].subject + " " + classes[i].catalogNumber + "\n")
+			fmt.Println("could not find course " + classes[i].subject + " " + classes[i].catalogNumber)
```

Comment:
To decrease the indentation, you can return if the error is nil instead of adding an else block.

Reply:
Since this is a loop, use a continue statement